src/swap/saga.ts

```diff
@@ -247,8 +250,182 @@ export function* swapSubmitSaga(action: PayloadAction<SwapInfo>) {
   }
 }
 
+export function* swapSubmitPreparedSaga(action: PayloadAction<SwapInfoPrepared>) {
+  const swapSubmittedAt = Date.now()
+  const {
+    price,
+    buyTokenAddress,
+    sellTokenAddress,
+    buyAmount,
+    sellAmount,
+    allowanceTarget,
+    estimatedPriceImpact,
+    guaranteedPrice,
+  } = action.payload.quote.rawSwapResponse.unvalidatedSwapTransaction
+  const amountType =
+    action.payload.userInput.updatedField === Field.TO
+      ? ('buyAmount' as const)
+      : ('sellAmount' as const)
+  const amount = action.payload.quote.rawSwapResponse.unvalidatedSwapTransaction[amountType]
+  const quoteReceivedAt = action.payload.quote.receivedAt
+
+  const tokens = yield* select((state) =>
+    tokensByIdSelector(state, [networkConfig.defaultNetworkId])
+  )
+  const fromToken =
+    tokens[getTokenId(networkConfig.defaultNetworkId, action.payload.userInput.fromToken)]
+  const toToken =
+    tokens[getTokenId(networkConfig.defaultNetworkId, action.payload.userInput.toToken)]
+
+  if (!fromToken || !toToken) {
+    Logger.error(
+      TAG,
+      `Could not find to or from token for swap from ${sellTokenAddress} to ${buyTokenAddress}`
+    )
+    yield* put(swapError())
+    return
+  }
+
+  const fromTokenBalance = fromToken.balance.shiftedBy(fromToken.decimals).toString()
+  const estimatedSellTokenUsdValue = calculateEstimatedUsdValue({
+    tokenInfo: fromToken,
+    tokenAmount: sellAmount,
+  })
+  const estimatedBuyTokenUsdValue = calculateEstimatedUsdValue({
+    tokenInfo: toToken,
+    tokenAmount: buyAmount,
+  })
+
+  const swapApproveContext = newTransactionContext(TAG, 'Swap/Approve')
+  const swapExecuteContext = newTransactionContext(TAG, 'Swap/Execute')
+
+  const defaultSwapExecuteProps = {
+    toToken: buyTokenAddress,
+    fromToken: sellTokenAddress,
+    amount,
+    amountType,
+    price,
+    allowanceTarget,
+    estimatedPriceImpact,
+    provider: action.payload.quote.rawSwapResponse.details.swapProvider,
+    fromTokenBalance,
+    swapExecuteTxId: swapExecuteContext.id,
+    swapApproveTxId: swapApproveContext.id,
+    estimatedSellTokenUsdValue,
+    estimatedBuyTokenUsdValue,
+  }
+
+  let quoteToTransactionElapsedTimeInMs: number | undefined
+
+  const getTimeMetrics = (): SwapTimeMetrics => ({
+    quoteToUserConfirmsSwapElapsedTimeInMs: swapSubmittedAt - quoteReceivedAt,
+    quoteToTransactionElapsedTimeInMs,
+  })
+
+  try {
+    // Navigate to swap pending screen
+    navigate(Screens.SwapExecuteScreen)
+
+    const wallet = yield* call(getViemWallet, networkConfig.viemChain.celo)
+    if (!wallet.account) {
+      // this should never happen
+      throw new Error('no account found in the wallet')
+    }
+
+    const preparedTransactions = action.payload.quote.preparedTransactions
+    if (preparedTransactions?.type !== 'possible') {
```

Comment:
would be nice if we had static type guarantees for this, instead of having to check it at runtime

Reply:
Yes it will eventually.